File: fsm-java/FSM/src/main/java/chat/dim/type/Duration.java
/* license: https://mit-license.org
 *
 *  Finite State Machine
 *
 *                                Written in 2024 by Moky <albert.moky@gmail.com>
 *
 * ==============================================================================
 * The MIT License (MIT)
 *
 * Copyright (c) 2024 Albert Moky
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 * ==============================================================================
 */
package chat.dim.type;

import java.io.Serializable;
import java.util.Date;

/**
 *  A span of time
 *  ~~~~~~~~~~~~~~
 *  such as 27 days, 4 hours, 12 minutes, and 3 seconds.
 */
public final class Duration implements Comparable<Duration>, Serializable {

    public static final Duration ZERO = new Duration(0);

    private final long millis;

    public Duration(long millis) {
        this.millis = millis;
    }

    public long inMilliseconds() {
        return millis;
    }

    /**
     * Compares this duration to the specified {@code Duration}.
     * <p>
     * The comparison is based on the total length of the durations.
     * It is "consistent with equals", as defined by {@link Comparable}.
     *
     * @param other  the other duration to compare to, not null
     * @return the comparator value, negative if less, positive if greater
     */
    @Override
    public int compareTo(Duration other) {
        return Long.compare(millis, other.inMilliseconds());
    }

    @Override
    public boolean equals(Object other) {
        return other instanceof Duration &&
                millis == ((Duration) other).inMilliseconds();
    }

    @Override
    public int hashCode() {
        return Long.hashCode(millis);
    }

    /**
     * A string representation of this duration using ISO-8601 seconds
     * based representation, such as {@code PT8H6M12.345S}.
     * <p>
     * The format of the returned string will be {@code PTnHnMnS}, where n is
     * the relevant hours, minutes or seconds part of the duration.
     * Any fractional seconds are placed after a decimal point i the seconds section.
     * If a section has a zero value, it is omitted.
     * The hours, minutes and seconds will all have the same sign.
     * <p>
     * Examples:
     * <pre>
     *    "20.345 seconds"                 -- "PT20.345S
     *    "15 minutes" (15 * 60 seconds)   -- "PT15M"
     *    "10 hours" (10 * 3600 seconds)   -- "PT10H"
     *    "2 days" (2 * 86400 seconds)     -- "PT48H"
     * </pre>
     * Note that multiples of 24 hours are not output as days to avoid confusion
     * with {@code Period}.
     *
     * @return an ISO-8601 representation of this duration, not null
     */
    @Override
    public String toString() {
        if (isZero()) {
            return "PT0S";
        }
        long hours = millis / MILLIS_PER_HOUR;
        long minutes = (millis % MILLIS_PER_HOUR) / MILLIS_PER_MINUTE;
        long secs = (millis % MILLIS_PER_MINUTE) / MILLIS_PER_SECOND;
        long ms = millis % MILLIS_PER_SECOND;
        StringBuilder buf = new StringBuilder();
        buf.append("PT");
        if (hours != 0) {
            buf.append(hours).append('H');
        }
        if (minutes != 0) {
            buf.append(minutes).append('M');
        }
        // seconds
        if (secs == 0 && ms == 0 && buf.length() > 2) {
            return buf.toString();
        } else if (secs >= 0 || ms <= 0) {
            buf.append(secs);
        } else if (secs == -1) {
            buf.append("-0");
        } else {
            buf.append(secs + 1);
        }
        // milliseconds
        if (ms > 0) {
            int pos = buf.length();
            if (secs < 0) {
                buf.append(2 * MILLIS_PER_SECOND - ms);
            } else {
                buf.append(ms + MILLIS_PER_SECOND);
            }
            while (buf.charAt(buf.length() - 1) == '0') {
                buf.setLength(buf.length() - 1);
            }
            buf.setCharAt(pos, '.');
        }
        buf.append('S');
        return buf.toString();
    }

    public boolean isZero() {
        return millis == 0;
    }

    public boolean isPositive() {
        return millis > 0;
    }

    public boolean isNegative() {
        return millis < 0;
    }

    public Duration negated() {
        return new Duration(-millis);
    }

    public Duration abs() {
        return isNegative() ? negated() : this;
    }

    //
    //  +
    //

    public Duration plus(Duration duration) {
        return plusMilliseconds(duration.inMilliseconds());
    }

    public Duration plusMilliseconds(long milliseconds) {
        return milliseconds == 0 ? this :
                new Duration(millis + milliseconds);
    }

    public Duration plusSeconds(long seconds) {
        return seconds == 0 ? this :
                new Duration(millis + MILLIS_PER_SECOND * seconds);
    }

    public Duration plusMinutes(int minutes) {
        return minutes == 0 ? this :
                new Duration(millis + MILLIS_PER_MINUTE * minutes);
    }

    public Duration plusHours(int hours) {
        return hours == 0 ? this :
                new Duration(millis + MILLIS_PER_HOUR * hours);
    }

    public Duration plusDays(int days) {
        return days == 0 ? this :
                new Duration(millis + MILLIS_PER_DAY * days);
    }

    //
    //  -
    //

    public Duration minus(Duration duration) {
        return plusMilliseconds(-duration.inMilliseconds());
    }

    public Duration minusMilliseconds(long milliseconds) {
        return plusMilliseconds(-milliseconds);
    }

    public Duration minusSeconds(long seconds) {
        return plusSeconds(-seconds);
    }

    public Duration minusMinutes(int minutes) {
        return plusMinutes(-minutes);
    }

    public Duration minusHours(int hours) {
        return plusHours(-hours);
    }

    public Duration minusDays(int days) {
        return plusDays(-days);
    }

    //
    //  *, /
    //

    public Duration multiply(long multiplicand) {
        if (multiplicand == 0) {
            return ZERO;
        } else if (multiplicand == 1) {
            return this;
        }
        return new Duration(millis * multiplicand);
    }

    public Duration divide(long divisor) {
        if (divisor == 0) {
            throw new ArithmeticException("Cannot divide by zero");
        } else if (divisor == 1) {
            return this;
        }
        return new Duration(millis / divisor);
    }

    //
    //  DateTime
    //

    public Date addTo(Date time) {
        return new Date(time.getTime() + millis);
    }

    public Date subtractFrom(Date time) {
        return new Date(time.getTime() - millis);
    }

    /**
     * Hours per day.
     */
    public static final int HOURS_PER_DAY = 24;
    /**
     * Minutes per hour.
     */
    public static final int MINUTES_PER_HOUR = 60;
    /**
     * Seconds per minute.
     */
    public static final int SECONDS_PER_MINUTE = 60;
    /**
     * Milliseconds per second.
     */
    public static final long MILLIS_PER_SECOND = 1000L;
    /**
     * Milliseconds per minute.
     */
    public static final long MILLIS_PER_MINUTE = MILLIS_PER_SECOND * SECONDS_PER_MINUTE;
    /**
     * Milliseconds per hour.
     */
    public static final long MILLIS_PER_HOUR = MILLIS_PER_MINUTE * MINUTES_PER_HOUR;
    /**
     * Milliseconds per day.
     */
    public static final long MILLIS_PER_DAY = MILLIS_PER_HOUR * HOURS_PER_DAY;

    //
    //  Factories
    //

    public static Duration ofMilliseconds(long milliseconds) {
        return new Duration(milliseconds);
    }

    public static Duration ofSeconds(long seconds) {
        return new Duration(MILLIS_PER_SECOND * seconds);
    }

    public static Duration ofMinutes(int minutes) {
        return new Duration(MILLIS_PER_MINUTE * minutes);
    }

    public static Duration ofHours(int hours) {
        return new Duration(MILLIS_PER_HOUR * hours);
    }

    public static Duration ofDays(int days) {
        return new Duration(MILLIS_PER_DAY * days);
    }

    /**
     *  The result of this method can be a negative period if the end is before the start.
     */
    public static Duration between(Date startTime, Date endTime) {
        return new Duration(endTime.getTime() - startTime.getTime());
    }

}

File: fsm-js/FSM/src/type/duration.js
'use strict';
// license: https://mit-license.org
// =============================================================================
// The MIT License (MIT)
//
// Copyright (c) 2025 Albert Moky
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.
// =============================================================================
//

//! require 'namespace.js'

    /// The number of milliseconds per second.
    var MILLISECONDS_PER_SECOND = 1000;

    /// The number of seconds per minute.
    ///
    /// Notice that some minutes of official clock time might
    /// differ in length because of leap seconds.
    /// The [Duration] and [DateTime] classes ignore leap seconds
    /// and consider all minutes to have 60 seconds.
    var SECONDS_PER_MINUTE      = 60;

    /// The number of minutes per hour.
    var MINUTES_PER_HOUR        = 60;

    /// The number of hours per day.
    ///
    /// Notice that some days may differ in length because
    /// of time zone changes due to daylight saving.
    /// The [Duration] class is time zone agnostic and
    /// considers all days to have 24 hours.
    var HOURS_PER_DAY           = 24;

    var MILLISECONDS_PER_MINUTE = MILLISECONDS_PER_SECOND * SECONDS_PER_MINUTE;
    var MILLISECONDS_PER_HOUR   = MILLISECONDS_PER_MINUTE * MINUTES_PER_HOUR;
    var MILLISECONDS_PER_DAY    = MILLISECONDS_PER_HOUR   * HOURS_PER_DAY;

    /**
     *  Time Duration
     */
    fsm.type.Duration = function (duration) {
        var days         = Converter.getInt(duration['days'],         0);
        var hours        = Converter.getInt(duration['hours'],        0);
        var minutes      = Converter.getInt(duration['minutes'],      0);
        var seconds      = Converter.getInt(duration['seconds'],      0);
        var milliseconds = Converter.getInt(duration['milliseconds'], 0);

        this.__millis = milliseconds +
            MILLISECONDS_PER_SECOND * seconds +
            MILLISECONDS_PER_MINUTE * minutes +
            MILLISECONDS_PER_HOUR   * hours +
            MILLISECONDS_PER_DAY    * days;
    };
    var Duration = fsm.type.Duration;

    /// The number of whole milliseconds spanned by this [Duration].
    ///
    /// The returned value can be greater than 999.
    /// For example, a duration of three seconds and 125 milliseconds
    /// has 3125 milliseconds.
    /// ```dart
    /// const duration = Duration(seconds: 3, milliseconds: 125);
    /// print(duration.inMilliseconds); // 3125
    /// ```
    Duration.prototype.inMilliseconds = function () {
        return this.__millis;
    };

    /// Whether this [Duration] has the same length as [other].
    ///
    /// Durations have the same length if they have the same number
    /// of milliseconds.
    Duration.prototype.equals = function (other) {
        return this.__millis === get_millis(other);
    };

    /// Whether this [Duration] is shorter than [other].
    Duration.prototype.shorterThan = function (other) {
        return this.__millis < get_millis(other);
    };

    /// Whether this [Duration] is longer than [other].
    Duration.prototype.longerThan = function (other) {
        return this.__millis > get_millis(other);
    };

    /// Whether this [Duration] is shorter than or equal to [other].
    Duration.prototype.shorterOrEqual = function (other) {
        return this.__millis <= get_millis(other);
    };

    /// Whether this [Duration] is longer than or equal to [other].
    Duration.prototype.longerOrEqual = function (other) {
        return this.__millis >= get_millis(other);
    };

    /// Creates a new [Duration] representing the absolute length of this
    /// [Duration].
    ///
    /// The returned [Duration] has the same length as this one, but is always
    /// positive where possible.
    Duration.prototype.abs = function () {
        return create_duration(Math.abs(this.__millis));
    };

    Duration.prototype.negated = function () {
        return create_duration(0 - this.__millis);
    };

    Duration.prototype.isNegative = function () {
        return this.__millis < 0;
    };

    Duration.prototype.isPositive = function () {
        return this.__millis > 0;
    };

    Duration.prototype.isZero = function () {
        return this.__millis === 0;
    };

    /// Adds this Duration and [other] and
    /// returns the sum as a new Duration object.
    Duration.prototype.adds = function (other) {
        return create_duration(this.__millis + get_millis(other));
    };

    /// Subtracts [other] from this Duration and
    /// returns the difference as a new Duration object.
    Duration.prototype.subtracts = function (other) {
        return create_duration(this.__millis - get_millis(other));
    };

    /// Multiplies this Duration by the given [factor] and returns the result
    /// as a new Duration object.
    ///
    /// Note that when [factor] is a double, and the duration is greater than
    /// 53 bits, precision is lost because of double-precision arithmetic.
    Duration.prototype.multiplies = function (factor) {
        return create_duration(Math.round(this.__millis * factor));
    };

    /// Divides this Duration by the given [quotient] and returns the truncated
    /// result as a new Duration object.
    ///
    /// The [quotient] must not be `0`.
    Duration.prototype.divides = function (quotient) {
        return create_duration(Math.floor(this.__millis / quotient));
    };

    var get_millis = function (duration) {
        if (Interface.conforms(duration, Duration)) {
            return duration.inMilliseconds();
        } else {
            return Converter.getInt(duration, 0);
        }
    };
    var create_duration = function (millis) {
        return new Duration({
            'milliseconds': millis
        });
    };

    //
    //  Factories
    //

    Duration.ofMilliseconds = function (millis) {
        return create_duration(millis * 1);
    };

    Duration.ofSeconds = function (seconds) {
        return create_duration(MILLISECONDS_PER_SECOND * seconds);
    };

    Duration.ofMinutes = function (minutes) {
        return create_duration(MILLISECONDS_PER_MINUTE * minutes);
    };

    Duration.ofHours = function (hours) {
        return create_duration(MILLISECONDS_PER_HOUR * hours);
    };

    Duration.ofDays = function (days) {
        return create_duration(MILLISECONDS_PER_DAY * days);
    };

    /**
     *  The result of this method can be a negative period if the end is before the start.
     *
     * @param {Date} startTime
     * @param {Date} endTime
     * @return {Duration}
     */
    Duration.between = function (startTime, endTime) {
        return create_duration(endTime.getTime() - startTime.getTime());
    };

    //
    //  DateTime
    //

    /**
     *  Adds this duration to a time
     *
     * @param {Date} time
     * @return {Date}
     */
    Duration.prototype.addTo = function (time) {
        return new Date(time.getTime() + this.__millis);
    };

    /**
     *  Subtracts this duration from a time
     *
     * @param {Date} time
     * @return {Date}
     */
    Duration.prototype.subtractFrom = function (time) {
        return new Date(time.getTime() - this.__millis);
    };
;